pkg/controller/direct/sql/sqlinstance_controller.go

```diff
@@ -138,7 +138,60 @@ func (a *sqlInstanceAdapter) Create(ctx context.Context, u *unstructured.Unstruc
 		return fmt.Errorf("resourceID is empty")
 	}
 
-	desiredGCP, err := SQLInstanceKRMToGCP(a.desired, a.refs)
+	if a.desired.Spec.CloneSource != nil {
+		return a.cloneInstance(ctx, u, log)
+	} else {
+		return a.insertInstance(ctx, u, log)
+	}
+}
+
+func (a *sqlInstanceAdapter) cloneInstance(ctx context.Context, u *unstructured.Unstructured, log klog.Logger) error {
+	desiredGCP, err := SQLInstanceKRMToGCPCloneRequest(a.desired, a.refs)
+	if err != nil {
+		return err
+	}
+
+	op, err := a.sqlInstancesClient.Clone(a.projectID, a.refs.sourceSQLInstance, desiredGCP).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("clone SQLInstance %s failed: %w", a.desired.Name, err)
+	}
+
+	pollingBackoff := gax.Backoff{
+		Initial:    time.Second,
+		Max:        time.Minute,
+		Multiplier: 2,
+	}
+	for {
+		log.V(2).Info("polling", "op", op)
+
+		if op.Status == "DONE" {
+			break
+		}
+		if err := gax.Sleep(ctx, pollingBackoff.Pause()); err != nil {
+			return fmt.Errorf("wait SQLInstance %s clone failed: %w", a.desired.Name, err)
```

Comment:
Nit: I personally think it reads better to use the "ing" form

i.e.

`failed to reconcile: waiting for SQLInstance %s clone: ...`

vs

`failed to reconcile: wait SQLInstance %s clone: ...`

Reply:
Fixed, and for all the other SQLInstance error messages also.